### PR TITLE
Pallet-Connectors-Gateway: Add Precompile for Axelar

### PR DESCRIPTION
**Pull request by mustermeiszer**
# Description
Adding a precompile.

@cdamian this is my first approach for the precompile. LMKWYT. 

cc @branan

### CODE REVIEW

**Comment by mustermeiszer** [comment]
@cdamian while doing that I noticed we need to nake the input of `process_msg` to a bounded `BoundedVec`. Just to note that here.

**Comment by mustermeiszer** [comment]
FYI @mikiquantum

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+	//
+	// Note: The _execute logic in this case will forward all calls to the
+	//       pallet-connectors-gateway with a special runtime local origin
+	//
+	#[precompile::public("execute(bytes32,string,string,bytes)")]
+	fn execute(
+		handle: &mut impl PrecompileHandle,
+		command_id: U256,
+		source_chain: U256,
+		source_address: U256,
```
Comment (cdamian) [inline pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:80]: This should be a `H160` right?

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	//     ) external {
+	//       ...
+	//     }
+	// Note: NOT SUPPORTED
+	//
+	#[precompile::public("forecallWithToken(address,uint256)")]
+	fn forecall_with_token(
+		handle: &mut impl PrecompileHandle,
+		source_chain: U256,
+		source_address: U256,
```
Comment (cdamian) [inline pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:127]: Same here - `H160`?

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	//     }
+	//
+	// Note: NOT SUPPORTED
+	//
+	#[precompile::public("executeWithToken(address,uint256)")]
+	fn execute_with_token(
+		handle: &mut impl PrecompileHandle,
+		command_id: U256,
+		source_chain: U256,
+		source_address: U256,
```
Comment (cdamian) [inline pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:157]: Same here - `H160`?

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+use precompile_utils::prelude::*;
+use sp_core::{ConstU32, Get, H160, U256};
+
+/// Precompile implementing IAxelarForecallable.
+/// MUST be used as the receiver of calls over the axelar bridge
+pub struct AxelarForecallable<Runtime, Gateway, MaxPayload>(
+	core::marker::PhantomData<(Runtime, Gateway, MaxPayload)>,
+);
+
+#[precompile_utils::precompile]
```
Comment (cdamian) [inline pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:26]: Do we need any other "EVM glue" for this to work or are these macro invocations enough to "register" this contract in the EVM?

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use fp_evm::PrecompileHandle;
+use frame_support::dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo};
+use pallet_connectors_gatway::GatwayOrigin;
```
Comment (cdamian) [inline pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:16]: Typo ^^

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	//       ...
+	//     }
+	//
+	// Note: NOT SUPPORTED
+	//
+	#[precompile::public("forecall(string,string,bytes,address)")]
+	fn forecall(
+		handle: &mut impl PrecompileHandle,
+		source_chain: U256,
+		sourceAddress: U256,
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:53]: Its a Solidity `string` which is represented as an `bytes32` also. IIRC. See `IAxelarForecallable.sol` in the repo

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs (hunk 2 above)
Comment (mustermeiszer) [reply pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:127]: Its a Solidity `string` which is represented as an `bytes32` also. IIRC. See `IAxelarForecallable.sol` in the repo

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs (hunk 3 above)
Comment (mustermeiszer) [reply pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:157]: Its a Solidity `string` which is represented as an `bytes32` also. IIRC. See `IAxelarForecallable.sol` in the repo

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs (hunk 4 above)
Comment (mustermeiszer) [reply pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:26]: They are enough to `impl Precompile` and some glue around generating selectors and so for the pallet. But we should expand it at least once ^^

File: pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs (hunk 5 above)
Comment (mustermeiszer) [reply pallets/connectors-gateway/connectors-gateway-axelar-precompile/src/lib.rs:16]: Haha. Always do that wrong. ^^

**Pull request merged**